qiskit/pulse/builder.py

```diff
@@ -0,0 +1,1996 @@
+        self._compile_lazy_circuit()
+        self._circuit_scheduler_settings = settings
+
+    @_compile_lazy_circuit_before
+    def compile(self) -> Schedule:
+        """Compile and output the built pulse program."""
+        # Not much happens because we currently compile as we build.
+        # This should be offloaded to a true compilation module
+        # once we define a more sophisticated IR.
+        program = self._schedule.append(self.block, mutate=True)
+        self.set_active_block(Schedule())
+        return program
+
+    @_compile_lazy_circuit_before
+    def set_active_block(self, block: Schedule):
+        """Set the active block for the builder."""
+        assert isinstance(block, Schedule)
+        self._block = block
+
+    @_compile_lazy_circuit_before
+    def append_block(self, block: Schedule):
+        """Add a block to the active schedule block.
+
+        Args:
+            block: Schedule to append to the active schedule block.
+        """
+        self.block.append(block, mutate=True)
+
+    @_compile_lazy_circuit_before
+    def append_instruction(self, instruction: instructions.Instruction):
+        """Add an instruction to the active schedule block.
+
+        Args:
+            instruction: Instruction to append.
+        """
+        self.block.append(instruction, mutate=True)
+
+    def _compile_lazy_circuit(self):
+        """Call a QuantumCircuit and append the output pulse schedule
+        to the active block."""
+        # check by length, can't check if QuantumCircuit is None
+        # so disable pylint error.
+        if self._lazy_circuit:
+            import qiskit.compiler as compiler  # pylint: disable=cyclic-import
+
+            lazy_circuit = self._lazy_circuit
+            # reset lazy circuit
+            self._lazy_circuit = self.new_circuit()
+            transpiled_circuit = compiler.transpile(lazy_circuit,
+                                                    self.backend,
+                                                    **self.transpiler_settings)
+            sched = compiler.schedule(transpiled_circuit,
+                                      self.backend,
+                                      **self.circuit_scheduler_settings)
+            self.call_schedule(sched)
+
+    def call_schedule(self, schedule: Schedule):
+        """Call a schedule and append to the active block."""
+        self.append_block(schedule)
+
+    def new_circuit(self):
+        """Create a new circuit for lazy circuit scheduling."""
+        return circuit.QuantumCircuit(self.num_qubits)
+
+    def _call_circuit(self, circ):
+        if self._lazy_circuit is None:
+            self._lazy_circuit = self.new_circuit()
+
+        for creg in circ.cregs:
+            try:
+                self._lazy_circuit.add_register(creg)
+            except circuit.exceptions.CircuitError:
+                pass
+        self._lazy_circuit.compose(circ, inplace=True)
```

Comment:
What does `inplace` do? Is `compose` necessary so that the input circuit won't be modified?

Reply:
`inplace` does a mutable compose of `_lazy_circuit`. `compose` is necessary because everything else complains if the circuits don't share the same registers on the lhs and rhs. This doesn't have anything to do with compilation, just extending the circuit.